### ISSUE TITLE
Optimize @check_output (pandera) when using with @extract_columns

### ISSUE THREAD

**Issue by IgorHoholko** [stitchfix/hamilton#301]
I want to use **pandera** `@check_output` togetrher with `@extract_columns` for dataloaders. After writing the code with compining both these decorators (see Fig below) I am getting some errors.
![image](https://user-images.githubusercontent.com/27057946/215851460-16843ab8-c7e0-4c44-9bb6-a014f3b88cee.png)

**The behaviour I would like to get:**
Validate only features I need at the moment. It means if from all `@extract_columns` I need only `spend` I would like to validate only `spend`.

**Code to reproduce:**

```
import logging
import sys

import numpy as np
import pandas as pd
import pandera as pa

from hamilton import ad_hoc_utils, driver
from hamilton.function_modifiers import extract_columns, check_output



schema = pa.DataFrameSchema({
    "signups": pa.Column(int, checks=pa.Check.le(10)),
    "spend": pa.Column(float, checks=pa.Check.lt(-1.2)),
    "other": pa.Column(str, checks=[
        pa.Check.str_startswith("value_"),
        # define custom checks as functions that take a series as input and
        # outputs a boolean or boolean Series
        pa.Check(lambda s: s.str.split("_", expand=True).shape[1] == 2)
    ]),
})



@extract_columns('signups', 'spend', 'other')
@check_output(schema = schema)
def log() -> pd.DataFrame:
    # data to validate
    df = pd.DataFrame({
        "signups": [1, 4, 0, 10, 12],
        "spend": [-1.3, -1.4, -2.9, -10.1, -20.4],
        "other": ["value_1", "value_2", "value_3", "value_2", "value1"],
    })
    return df




# # Look at `my_functions` to see how these functions connect.
# def signups() -> pd.Series:
#     """Returns sign up values"""
#     return pd.Series([1, 10, 50, 100, 200, 400])


# def spend() -> pd.Series:
#     """Returns the spend values"""
#     return pd.Series([10, 10, 20, 40, 40, 50])


def log_spend_per_signup(spend_per_signup: pd.Series) -> pd.Series:
    """Simple function taking the logarithm of spend over signups."""
    return np.log(spend_per_signup)


# Place the functions into a temporary module -- the idea is that this should house a curated set of functions.
# Don't be afraid to make multiple of them -- however we'd advise you to not use this method for production.
# Also note, that using a temporary function module does not work for scaling onto Ray, Dask, or Pandas on Spark.
temp_module = ad_hoc_utils.create_temporary_module(
    log, log_spend_per_signup, module_name="function_example"
)







import pandas as pd

"""
Notes:
  1. This file is used for all the [ray|dask|spark]/hello_world examples.
  2. It therefore show cases how you can write something once and not only scale it, but port it
     to different frameworks with ease!
"""


def avg_3wk_spend(spend: pd.Series) -> pd.Series:
    """Rolling 3 week average spend."""
    return spend.rolling(3).mean()


def spend_per_signup(spend: pd.Series, signups: pd.Series) -> pd.Series:
    """The cost per signup in relation to spend."""
    return spend / signups


def spend_mean(spend: pd.Series) -> float:
    """Shows function creating a scalar. In this case it computes the mean of the entire column."""
    return spend.mean()


def spend_zero_mean(spend: pd.Series, spend_mean: float) -> pd.Series:
    """Shows function that takes a scalar. In this case to zero mean spend."""
    return spend - spend_mean


def spend_std_dev(spend: pd.Series) -> float:
    """Function that computes the standard deviation of the spend column."""
    return spend.std()


def spend_zero_mean_unit_variance(spend_zero_mean: pd.Series, spend_std_dev: float) -> pd.Series:
    """Function showing one way to make spend have zero mean and unit variance."""
    return spend_zero_mean / spend_std_dev


my_functions = ad_hoc_utils.create_temporary_module(
    avg_3wk_spend, spend_per_signup, spend_mean, spend_zero_mean, 
    spend_std_dev, spend_zero_mean_unit_variance, module_name="my_functions"
)







# Cell 4 - Instantiate the Hamilton driver and pass it the right things in.

initial_config = {}
# we need to tell hamilton where to load function definitions from
dr = driver.Driver(initial_config, my_functions, temp_module)  # can pass in multiple modules
# we need to specify what we want in the final dataframe.
output_columns = [
    "spend",
    "signups",
    "avg_3wk_spend",
    "spend_per_signup",
    "spend_zero_mean_unit_variance",
    'other'
]
# Cell 5 - visualize execution
# To visualize do `pip install "sf-hamilton[visualization]"` if you want these to work





dr.execute(output_columns)
```

**Comment by skrawcz** [comment]
@IgorHoholko just following up. Thanks for the submission. I'll have to scope this and see whether we can give you the API that you want. It doesn't sound unreasonable, we just need ensure we can do it in a modular way that makes sense.

**Comment by elijahbenizzy** [comment]
Chiming in -- yep, definitely possible. We'll have to be a little clever, but I don't see why the current abstractions wouldn't work for this -- we'll just need to change things slightly.

**Comment by skrawcz** [comment]
Yeah so we need to rethink things a little here with the data validators/decorator ordering and how data validators are selected... so not as easy of a fix as it should be. i.e. we need to validate the type of the object passed in to the schema kwarg, and have check_output apply to the right node, rather than to the final set of nodes... but that brings up a bunch of edge cases.

**Comment by skrawcz** [comment]
Specifically:

**Issue 1**
In this example, the pandera schema is a dataframe schema. But the datavalidator created is the one for Series. We should prevent this from happening by checking that the class passed in matches the expected class of the validator. Perhaps singledispatch would be useful here? Though I'm not entirely sure.

**Issue 2**
From a UX perspective, one would expect to apply the check_output function before the extract_columns function. Instead check_output is applied after extract_columns (and thus why issue 1 resolves the way it does now).

Questions this brings up:
 - do we have the right terminology and classifications in terms of "node operations", i.e. expand, transform, etc., for each decorator.
 - why don't we just use decorator order as the way to resolve nodes? I don't recall the decision for that.

**Comment by elijahbenizzy** [comment]
> Specifically:
> 
> **Issue 1** In this example, the pandera schema is a dataframe schema. But the datavalidator created is the one for Series. We should prevent this from happening by checking that the class passed in matches the expected class of the validator. Perhaps singledispatch would be useful here? Though I'm not entirely sure.
> 
> **Issue 2** From a UX perspective, one would expect to apply the check_output function before the extract_columns function. Instead check_output is applied after extract_columns (and thus why issue 1 resolves the way it does now).
> 
> Questions this brings up:
> 
> * do we have the right terminology and classifications in terms of "node operations", i.e. expand, transform, etc., for each decorator.
> * why don't we just use decorator order as the way to resolve nodes? I don't recall the decision for that.
> 
> **Issue 3** We need to handle the case where the Pandera DF schema is the superset of what is produced. Thankfully, Pandera allows for missing columns in schemas, and so we will be able to provide this type of functionality without having to do much. That is, the user will just need to set `required=False` for each column in the dataframe schema.

OK, so the data validator is a [NodeTransformer](https://github.com/stitchfix/hamilton/blob/7fe662a1f60ab4c632696e0b2d60c732fc3c7a76/hamilton/function_modifiers/base.py#L236), which means that it applies to all the "sinks" of a subdag. E.G. in the extract_columns case, it applies to all the series that are produced. I think we should keep this as a default behavior, but add a parameter to show *which* one this applies to. Calling it `target` below, but open to other names. I think this solves (1) and (2) at the same time, as well as allows integration with a host of other decorators.

```python
schema = pa.DataFrameSchema({
    "signups": pa.Column(int, checks=pa.Check.le(10)),
    "other": pa.Column(str, checks=[
        pa.Check.str_startswith("value_"),
        # define custom checks as functions that take a series as input and
        # outputs a boolean or boolean Series
        pa.Check(lambda s: s.str.split("_", expand=True).shape[1] == 2)
    ]),
})



@extract_columns('signups', 'spend', 'other')
@check_output(schema = schema, target='log') # Just applies to log
@check_output(schema=pa.SeriesSchema(float, checks=pa.Check.lt(-1.2)), target='spend') # Just applies to `spend`
@check_output(schema=pa.SeriesSchema(float, checks=[]) # applies to everything
def log() -> pd.DataFrame:
    # data to validate
    df = pd.DataFrame({
        "signups": [1, 4, 0, 10, 12],
        "spend": [-1.3, -1.4, -2.9, -10.1, -20.4],
        "other": ["value_1", "value_2", "value_3", "value_2", "value1"],
    })
    return df
```

As to how this is implemented, we need to be smarter about it -- basically its *currently* a `NodeTransformer`, which isn't powerful enough. We need this to be a [`SubDAGModifier`](https://github.com/stitchfix/hamilton/blob/7fe662a1f60ab4c632696e0b2d60c732fc3c7a76/hamilton/function_modifiers/base.py#L179), which allows it to do anything it wants with a collection of nodes. We could also add another abstraction `TargetedNodeModifier`, which would give it just the set of nodes it wants to modify, and then change the upstream dependencies to handle it appropriately.

In our work we're planning on adapters, these might both be valuable.

Re (3) I agree, that's on the user.

**Comment by skrawcz** [comment]
TL;DR:
 - "target" parameter can make sense, but is there anything generalizable about it?  or does it only apply to `check_output`? 
 - I think our decorator application/resolution logic can be simpler/clearer

What bothers me is that the order of decorators doesn't matter -- and the only way a user knows what is going on, is by having to read and understand each decorator's precedence; this seems like we're increasing the cognitive burden here? E.g. adding the "target" parameter for Igor's example is a reaction IMO to a decision the framework made, rather than ease of use, unless we require a `target` for every decorator? But then that seems like too much verbosity for 80% of use cases... 

 Otherwise nit, you'd either check `log` or what is `extracted`, not both IRL (I can't think of a situation where a decorator would change the output)?

For example -- your example below:
```python
@extract_columns('signups', 'spend', 'other')
@check_output(schema = schema, target='log') # Just applies to the actual dataframe output
@check_output(schema=pa.SeriesSchema(float, checks=pa.Check.lt(-1.2)), target='spend') # Just applies to `spend`
@check_output(schema=pa.SeriesSchema(float, checks=[]) # applies to everything
def log() -> pd.DataFrame:
    ...
    return df
```
should be the reverse, because my prior background tells me that with decorators you work your way up from the function to understand what's going on, i.e. we create a dataframe, can apply checks to it, then extract outputs, and then apply checks to those outputs.
```python
@check_output(schema=pa.SeriesSchema(float, checks=pa.Check.lt(-1.2)), target='spend') # Just applies to `spend`
@check_output(schema=pa.SeriesSchema(float, checks=[]) # applies to everything
@extract_columns('signups', 'spend', 'other')
@check_output(schema = schema) # target not required; Just applies to the actual dataframe output
def log() -> pd.DataFrame:
    ...
    return df
```
Alternatively we could enforce top down reading/application of decorators:
```python
@check_output(schema = schema) # Just applies to the actual dataframe output
@extract_columns('signups', 'spend', 'other')
@check_output(schema=pa.SeriesSchema(float, checks=pa.Check.lt(-1.2)), target='spend') # Just applies to `spend`
@check_output(schema=pa.SeriesSchema(float, checks=[]) # applies to everything
def log() -> pd.DataFrame:
    ...
    return df
```

@IgorHoholko feel free to chime in here.

**Comment by elijahbenizzy** [comment]
> TL;DR:
> 
> * "target" parameter can make sense, but is there anything generalizable about it?  or does it only apply to `check_output`?
> * I think our decorator application/resolution logic can be simpler/clearer
> 
> What bothers me is that the order of decorators doesn't matter -- and the only way a user knows what is going on, is by having to read and understand each decorator's precedence; this seems like we're increasing the cognitive burden here? E.g. adding the "target" parameter for Igor's example is a reaction IMO to a decision the framework made, rather than ease of use, unless we require a `target` for every decorator? But then that seems like too much verbosity for 80% of use cases...
> 
> Otherwise nit, you'd either check `log` or what is `extracted`, not both IRL (I can't think of a situation where a decorator would change the output)?
> 
> For example -- your example below:
> 
> ```python
> @extract_columns('signups', 'spend', 'other')
> @check_output(schema = schema, target='log') # Just applies to the actual dataframe output
> @check_output(schema=pa.SeriesSchema(float, checks=pa.Check.lt(-1.2)), target='spend') # Just applies to `spend`
> @check_output(schema=pa.SeriesSchema(float, checks=[]) # applies to everything
> def log() -> pd.DataFrame:
>     ...
>     return df
> ```
> 
> should be the reverse, because my prior background tells me that with decorators you work your way up from the function to understand what's going on, i.e. we create a dataframe, can apply checks to it, then extract outputs, and then apply checks to those outputs.
> 
> ```python
> @check_output(schema=pa.SeriesSchema(float, checks=[]) # applies to everything
> @check_output(schema=pa.SeriesSchema(float, checks=pa.Check.lt(-1.2)), target='spend') # Just applies to `spend`
> @extract_columns('signups', 'spend', 'other')
> @check_output(schema = schema) # target not required; Just applies to the actual dataframe output
> def log() -> pd.DataFrame:
>     ...
>     return df
> ```
> 
> Alternatively we could enforce top down reading/application of decorators:
> 
> ```python
> @check_output(schema = schema) # Just applies to the actual dataframe output
> @extract_columns('signups', 'spend', 'other')
> @check_output(schema=pa.SeriesSchema(float, checks=pa.Check.lt(-1.2)), target='spend') # Just applies to `spend`
> @check_output(schema=pa.SeriesSchema(float, checks=[]) # applies to everything
> def log() -> pd.DataFrame:
>     ...
>     return df
> ```
> 
> @IgorHoholko feel free to chime in here.

Ok, lots to think through here:

1. Decorator ordering is messy, but I'm not sure if we should go by "natural" order (E.G. the operations and what order they *can* go in, like we're doing now), or "presented order" (E.G. the operations and what order they're written in), or *reverse presented order* (in the other direction, probably not the wrong one, but decorators are unintuitive). Decorators being unintuitive led me to ensure that order shouldn't matter, but I get how that's not intuitive as well... I think the devil is in the documentation here. 

2. `target` only makes sense for a subset of decorators -- E.G. the ones that perform an operation on a single node. Those should be a subclass (currently, its `NodeTransformer`, of which DataValidator is the only implemented, but I think its pretty general), with sensible defaults if no target is supplied. A quick redefinition (filter, etc...) could be a decorator as well.
 
3. Yes, the example doesn't make sense logically, more just to demonstrate the capability of the API. That said, I think it *might* make sense to have two separate decorators with two targets in that case. A custom validator on the output of a column, then a pandera schema... Or something like that. Either way, no reason it should be impossible, just not the cannonical example.

**Comment by skrawcz** [comment]
Yup. Yeah so I think let's move forward with the `target` kwarg then at least to unblock this issue; I think a `target` option makes sense independent of decorator ordering at least here.

**Comment by skrawcz** [comment]
@IgorHoholko we just merged an update to enable `target_=` on the decorator. See #316.

We have this up as a release candidate `pip install sf-hamilton==1.17.0rc0` -- will likely release tomorrow assuming testing goes well.

Thanks for finding the issue!
